docs: Community Snippets

munsterlander opened the pull request:
# Description

Based on our conversation in discord, I have created the initial concept of recipes for the docs.  What they end up looking like could be completely different than this.  I know @spydon was thinking of embedding functioning games in this and maybe we can, but looking at Klondike, it seems that we would be committing a lot of files when people really just want the code snippet.  So maybe this isn't a recipe per se, but more of a snippet library.  None the less, as this is OSS, I felt it important to give credit back to the folks I took the initial tidbits from, so please feel free to adjust as necessary.

## Checklist

<!-- Before you create this PR confirm that it meets all requirements listed below by checking the
relevant checkboxes (`[x]`). This will ensure a smooth and quick review process. -->

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [ ] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [ ] I have updated/added relevant examples in `examples`.

## Breaking Change

<!-- Does your PR require Flame users to manually update their apps to accommodate your change? 

If the PR is a breaking change this should be indicated with suffix "!" 
(for example, `feat!:`, `fix!:`). See [Conventional Commit] for details.
-->

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

<!-- ### Migration instructions

If the PR is breaking, uncomment this header and add instructions for how to migrate from the
currently released version to the new proposed way.
-->

## Related Issues

<!-- Provide a list of issues related to this PR from the [issue database].
Indicate which of these issues are resolved or fixed by this PR, i.e. Fixes #xxxx* !-->

<!-- Links -->
[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md
[Conventional Commit]: https://conventionalcommits.org

spydon commented on doc/recipes/flavor_registry/audio.md:
```diff
@@ -0,0 +1,17 @@
+import 'package:audioplayers/audioplayers.dart';
+AudioPlayer? loopPlayer;
+  
+Future<void>startSound()async{
```
Seems like some spaces are missing here

munsterlander replied:
Spaces added.

spydon commented on doc/recipes/recipes.md:
```diff
@@ -0,0 +1,17 @@
+# Recipies
+
+Recipes are broken down into their common flavors below.  A specific recipe can be in multiple flavors depending upon the components utilized or design pattern contained within.  The following contains the list of tags utilized to categorize the recipies:
```
Max 100 char per line

munsterlander replied:
Done.

spydon reviewed (commented):
Good job! What I meant with embedded running code was more like how the router docs are, not full games like Klondike.

munsterlander commented:
> Good job! What I meant with embedded running code was more like how the router docs are, not full games like Klondike.

So I was looking at this but I didn't see it in router - I was reviewing the code in Decorators.  Router is what I wanted to do as well.  Let me see if I can figure out how these can all be one project - like Decorators and that they would be easy to add.  What I didn't know, was for the WASD keyboard example, would the event passthrough or be captured by the docs, etc.

spydon commented:
@munsterlander check here: https://docs.flame-engine.org/main/development/documentation.html#building-documentation-locally

munsterlander commented:
> @munsterlander check here: https://docs.flame-engine.org/main/development/documentation.html#building-documentation-locally

Well how did I not see that, lol!

munsterlander commented:
I don't know if its my environment or what (I have both Python 2&3 loaded due to work stuff), but I can't get it to make:

```console
Running Sphinx v5.1.1
myst v0.16.1: MdParserConfig(renderer='sphinx', commonmark_only=False, enable_extensions=['dollarmath', 'html_admonition', 'html_image', 'linkify', 'replacements', 'smartquotes'], linkify_fuzzy_links=True, dmath_allow_labels=True, dmath_allow_space=True, dmath_allow_digits=True, dmath_double_inline=False, update_mathjax=True, mathjax_classes='tex2jax_process|mathjax_process|math|output_area', disable_syntax=[], url_schemes=['http', 'https', 'mailto', 'ftp'], heading_anchors=4, heading_slug_func=None, html_meta=[], footnote_transition=True, substitutions=[], sub_delimiters=['{', '}'], words_per_minute=200)
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 55 source files that are out of date
updating environment: [new config] 55 added, 0 changed, 0 removed
Compiling Flutter app [flame-examples]documentation

Exception occurred:
  File "c:\users\munsterlander\appdata\local\programs\python\python36-32\lib\subprocess.py", line 997, in _execute_child
    startupinfo)
FileNotFoundError: [WinError 2] The system cannot find the file specified
The full traceback has been saved in C:\Users\munsterlander\AppData\Local\Temp\sphinx-err-3nye4z0r.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
```

st-pasha commented:
Ah, I guess we haven't tried building on a windows machine before :)

Would the command `flutter build web` work for you? Or does it have to be `flutter.exe build web`? Sorry, I never worked with Flutter on windows.

If it's the latter, then open the file `doc/_sphinx/extensions/flutter_app.py`, and in line 182 change `'flutter'` into `'flutter.exe'`. If after this the error disappears, then we'll figure out how to modify it to work properly on both windows and linux.

munsterlander commented:
`flutter build web` works fine.   The following gets generated in a folder that doesn't have a pubspec of course:

```console
flutter build web
Error: No pubspec.yaml file found.
This command should be run from the root of your Flutter project.
```

st-pasha commented:
So, what happens if you:
1. Go to the directory with doc examples: `cd doc/flame/examples`
2. Run python: `python`
3. In python console run command `import subprocess; subprocess.run(['flutter', 'build', 'web'])` ?

munsterlander commented:
> 3\. import subprocess; subprocess.run(['flutter', 'build', 'web'])

```console
Python 3.6.5 (v3.6.5:f59c0932b4, Mar 28 2018, 16:07:46) [MSC v.1900 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import subprocess; subprocess.run(['flutter', 'build', 'web'])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\munsterlander\AppData\Local\Programs\Python\Python36-32\lib\subprocess.py", line 403, in run
    with Popen(*popenargs, **kwargs) as process:
  File "C:\Users\munsterlander\AppData\Local\Programs\Python\Python36-32\lib\subprocess.py", line 709, in __init__
    restore_signals, start_new_session)
  File "C:\Users\munsterlander\AppData\Local\Programs\Python\Python36-32\lib\subprocess.py", line 997, in _execute_child
    startupinfo)
FileNotFoundError: [WinError 2] The system cannot find the file specified
>>> import subprocess;
>>> subprocess.run(['flutter', 'build', 'web']);
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\munsterlander\AppData\Local\Programs\Python\Python36-32\lib\subprocess.py", line 403, in run
    with Popen(*popenargs, **kwargs) as process:
  File "C:\Users\munsterlander\AppData\Local\Programs\Python\Python36-32\lib\subprocess.py", line 709, in __init__
    restore_signals, start_new_session)
  File "C:\Users\munsterlander\AppData\Local\Programs\Python\Python36-32\lib\subprocess.py", line 997, in _execute_child
    startupinfo)
FileNotFoundError: [WinError 2] The system cannot find the file specified
>>>
```

st-pasha commented:
What if you say `subprocess.run("flutter build web")`, or `subprocess.run("flutter build web", shell=True)`?

munsterlander commented:
> subprocess.run("flutter build web", shell=True)

That runs! Its currently compiling.

munsterlander commented:
New problem, but I should be able to solve it shortly.  The docs for the python http server apparently are incorrect:

```console
C:\Users\munsterlander\AndroidStudioProjects\flame>python -m http.server 8000 --directory doc/_build/html
usage: server.py [-h] [--cgi] [--bind ADDRESS] [port]
server.py: error: unrecognized arguments: --directory doc/_build/html
```

munsterlander commented:
@st-pasha So the whole issue was not the python code, but the python version.  What clued me in, was this comment about python version for the error regarding `dirs_exist_ok`:  https://stackoverflow.com/questions/63661976/problems-using-shutil-copytree-in-python

Anyway, once I upgraded to Python 3.8+ (I am running 3.9.x), I reverted the code and it all ran perfectly.  I will update the docs to state that Python 3.8+ is required and I would suspect there needs to be a check in the `requirements.txt` file as well.

st-pasha commented:
> I reverted the code and it all ran perfectly. 

So `subprocess.run("flutter build web", shell=True)` is not actually needed, it runs fine as-is? In that case bumping the min required version of Python would definitely be the easiest solution, especially since Python 3.6 has reached its end of life already: https://endoflife.date/python

munsterlander commented:
> > I reverted the code and it all ran perfectly.
> 
> So `subprocess.run("flutter build web", shell=True)` is not actually needed, it runs fine as-is? In that case bumping the min required version of Python would definitely be the easiest solution, especially since Python 3.6 has reached its end of life already: https://endoflife.date/python

Yes,  the original subprocess code runs fine. It was only the min version of python. I have updated the docs accordingly.

munsterlander commented:
Ok, now no clue what is going on.  The errors have come back.  I had to revert to the new subprocess command to get it building but then tons of errors regarding missing yaml files.  I will keep working on this.

st-pasha commented on doc/_sphinx/extensions/flutter_app.py:
```diff
@@ -178,13 +178,19 @@ def _ensure_compiled(self):
         try:
-            subprocess.run(
+            """ subprocess.run(
                 ['flutter', 'build', 'web'],
                 stdout=subprocess.PIPE,
                 stderr=subprocess.STDOUT,
                 cwd=self.source_dir,
                 check=True,
-            )
+            ) """
+            cwd = os.getcwd()
+            # Print the current working directory
+            print("Current working directory: {0}".format(cwd))
+            abspath = os.path.abspath("../flame/examples")
+            os.chdir(abspath)
+            subprocess.run("flutter build web", shell=True)
```
Try
```python
            subprocess.run(
                "flutter build web",
                shell=True,
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
                cwd=self.source_dir,
                check=True,
            )
```

munsterlander replied:
This worked as well.  Do you think we still need to change directory?  I commented it out and it still ran, but I had issues earlier about not finding the `pubspec.yaml` file for flutter build. 

This is what I currently have and it seems to be working ok:

```python
def _compile_source(self):
        try:
            subprocess.run(
                "flutter build web",
                shell=True,
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
                cwd=self.source_dir,
                check=True,
            )
        except subprocess.CalledProcessError as e:
            cmd = ' '.join(e.cmd)
            raise self.error(
                f'Command `{cmd}` returned with exit status {e.returncode}\n' +
                e.output.decode('utf-8'),
            )
```

st-pasha replied:
The `cwd` argument means "current working directory", so it should take care of that.

munsterlander replied:
Ah, I didn't even read the arguments. Good to know.

munsterlander commented:
Well thats odd....not certain why the build is failing.  Error is unable to install the flame package.  Nothing was changed to pubspecs after the last successful build.

st-pasha commented:
Note: there's file `doc/_sphinx/extensions/__pycache__/flutter_app.cpython-39.pyc` in the PR currently -- it's a temporary file that shouldn't be checked in. Perhaps add `pyc` extension to the gitignore?

munsterlander commented:
> Note: there's file `doc/_sphinx/extensions/__pycache__/flutter_app.cpython-39.pyc` in the PR currently -- it's a temporary file that shouldn't be checked in. Perhaps add `pyc` extension to the gitignore?

Will do and revert and remove.

spydon commented:
> I didn't change anything in the pubspec and the last person to change it was Spydon 11 days ago but this was building after that.  The only change is in the python, but that code runs fine locally and shouldn't have anything to do with the `flutter pub get` in `examples\examples` that is causing the failure.  I do not see these packages in the examples pubspec, nor where they ever there, so I assume they get injected from upstream.

Haven't had a look yet but I would guess it is due to the new Flutter version (or possibly some updates package).

spydon commented on doc/recipes/flavor_registry/controls.md:
````diff
@@ -0,0 +1,91 @@
+
+## Touch Controls within the HUD
+
+```dart
````
Is this to be embedded as a running example too?
Could we reference the dart file code for that example instead of having it in the markdown then maybe? Since if anything changes in Flame these snippets easily become outdated.

munsterlander replied:
Yes.  I was going to get around to doing this as an embed as well.  Just wanted to see your thoughts on what I had so far and direction.

munsterlander replied:
I just need to add this embed.  I will get to this tonight.

st-pasha replied:
Do let me know if you think we can augment the `flutter_app` extension in any way, to accommodate the recipes: adding a new mode, or changing the style, etc.

munsterlander replied:
So far, everything just works really well. I haven't found anything that I wasn't able to do. I will take a closer look at the parent and see if there is anything I think could be added. I will say it's been pretty cool because it just works!

munsterlander commented on doc/recipes/recipes.md:
````diff
@@ -0,0 +1,19 @@
+```{toctree}
+:hidden:
+
+Controls    <flavor_registry/controls.md>
````
No, I was just trying to keep things organized.  I was afraid if the decorators grew then its just a mess of files all named like `decorator_x.dart` and `recipe_x.dart` which can become unmanageable if there were 100's of files.

spydon replied:
We can change it if that becomes the case

munsterlander replied:
Removed

munsterlander commented on doc/flame/examples/lib/recipe_keyboard_wasd.dart:
```diff
@@ -0,0 +1,59 @@
+import 'package:flame/game.dart';
+import 'package:flutter/services.dart';
+
+class RecipeKeyboardWASDGame extends FlameGame
```
Can do.  I was just following the Decorator naming convention.

spydon replied:
Okay, maybe best to keep the filenames as is then

munsterlander commented on doc/flame/examples/lib/recipe_keyboard_wasd.dart:
```diff
@@ -0,0 +1,59 @@
+  }
+
+  @override
+  bool onKeyEvent(RawKeyEvent event, Set<LogicalKeyboardKey> keysPressed) {
```
I didn't realize that was in examples.  If that's the case, then this recipe should be dropped.  Recipes should just be for common patterns / snippets that do not need the full example treatment, imo.

munsterlander replied:
Removed.

munsterlander commented on doc/recipes/recipes.md:
```diff
@@ -0,0 +1,19 @@
+- [](flavor_registry/keyboard.md) -- Examples for basic keyboard interaction with a game or class
+- [](flavor_registry/audio.md) -- Examples for basic audio management within a game or class
+- [](flavor_registry/physics.md) -- Examples for basic physics interactions with sprites and other components
+
```
Yes.  I think this should really say snippet and not examples.  I wouldn't want to cause confusion.  Maybe recipes should be snippets.  Thoughts?

spydon replied:
I like snippets! That's a very clear name for what it is.

munsterlander replied:
Ok, let me update all of this to snippets.

st-pasha replied:
:+1: snippets

munsterlander replied:
Done.

st-pasha commented on doc/recipes/flavor_registry/audio.md:
````diff
@@ -0,0 +1,18 @@
+    loopPlayer!.stop();
+}
+```
+- Credit @viktorManual from [Discord Flame Channel](https://discord.com/channels/509714518008528896/516639688581316629/1012562750725169173)
````
@spydon Do we want to include attribution links like this? I'm worried that it could create barriers against improving the recipes. Like, if the recipe is tagged as "belonging to user X", then people could hesitate to change it in any significant way.

spydon replied:
True true, it would quickly become hard to maintain.
Better to skip these ones and just ask people if code is copied from somewhere else into the recipes.

munsterlander replied:
Not certain if it would prevent people from updating it, but as they were snippets, I was wanting to provide context if there was any as to why it was written the way it was and what the person was solving.  Additionally, I was hoping it would encourage people to contribute to the community.

st-pasha replied:
It is better to explain the context directly next to the snippet, so that the reader wouldn't be forced to visit another site and dig through comments. Plus, I'm not even sure how stable the Discord comment links are.

> Additionally, I was hoping it would encourage people to contribute to the community.

Indeed, this does provide some degree of encouragement. But at the same time, it also serves as discouragement for anyone to modify/improve existing snippets.

munsterlander replied:
Ok, let me rework the blocks a bit.  I updated the section to be called Community Snippets but it can be just Snippets if we are pulling any mentions or links.  Additionally, this all started fit me because I felt the examples were lacking, but wow have they been rounded out.  I'm starting to question the need for snippets all together. The one thing that I really like about them though, is embedding the app in the docs. It makes it real nice and clean.  Having to the leave the docs to go to examples and then having to open another window for code is tedious, so maybe there is room for snippets. 

Up to you guys though.

spydon replied:
Maybe there is no need for a separate section for snippets? But we could have live examples everywhere in the docs where we are describing the features. Quite a big job though because I think most examples on the examples page are made for bigger screens.

Maybe there are cases where the snippets are more advanced than just involving one feature though, not sure?

munsterlander replied:
I really like the embeds.  The examples (amazing how many there are now) are too far away from the docs imo.  I consistently see you giving the examples link in discord and people are like, oh, I didn't know that was there.  So either every section in the docs needs a link to examples or there is an embed (huge task and may not be possible for every area) or there are snippets.

I do think there is a need for snippets (and my original intention) for common patterns that aren't exactly an explanation of Flame, but more for common patterns.  Like the how do I stop audio or how do I arc a bullet, etc.  It could even be how do I store my level data or such.

spydon replied:
>... or there is an embed (huge task and may not be possible for every area) or there are snippets.

I guess this will be the goal in the long run.

> I do think there is a need for snippets (and my original intention) for common patterns that aren't exactly an explanation of Flame, but more for common patterns.  Like the how do I stop audio or how do I arc a bullet, etc.  It could even be how do I store my level data or such.

Sounds good!

st-pasha replied:
I like the idea of having a small example embed in the docs for every Flame feature or component. This is similar to how Flutter has video examples for every widget -- only better, because we'll have a real live code.

I also agree with @munsterlander that recipes/snippets would be a separate form of documentation -- focused not on individual Flame features, but rather on how they fit together, or how they can be used to achieve specific game-development-related goals. These should ideally also feature embeds (where feasible).

The third kind of documentation is the Tutorials -- these would focus on complete games, or maybe just some large-ish pieces of functionality.

What's a bit unclear to me is how the current Examples would fit into the picture. Should it be a separate category, or is it better to split them across the 3 kinds of docs mentioned above?

munsterlander replied:
Well, I really like the examples, but I don't like how far apart they are from the docs.   Same with the api.  It would be awesome that when you are in the flame docs looking at say joystick component, you have the embed as well a links to the example and the api docs.

I think there is room for everything but they need to be connected so it's cohesive. Like you were saying about flutter docs.  It's one place for everything.

munsterlander replied:
Ok, now the docs redo and the tutorials and such have been sorted, I want to come back to this.  I really think it should be community snippets and like @st-pasha is saying, it should be:

 _focused not on individual Flame features, but rather on how they fit together, or how they can be used to achieve specific game-development-related goals. These should ideally also feature embeds (where feasible)._

So with that in mind, I am going to scrap a lot of what was in this PR and redirect it to this goal.  This would still leave tutorials and examples completely intact, but give a small quick area for short abstract concepts using Flame or other gaming concepts.

munsterlander commented:
Ok don't know why it closed.  I accepted all incoming changes and just updated my branch.  Regardless, this PR was so far behind, I will open a new one and just reference this one.  I fear something may have got lost in the merge.
